### ISSUE TITLE
Double task execution on multi-machine setup

### ISSUE THREAD

**Issue by jtv** [Koed00/django-q#234]
This is a different issue from #183.  I didn't set a timeout, and setting it to 60 doesn't help.

I'm using Django-Q with the ORM broker.  I run two clusters on separate machines, both talking to the same PostgreSQL database, no caching.  When my application fires off a task, it gets executed twice.

Both executions happen in the same second.  The admin UI only shows a single completed task.

My suspicion is that task locking isn't working properly so that both clusters pick up the same task.  When I shut down one of my Django-Q clusters, so that I'm running only one, this no longer happens.  With a single cluster, tasks execute only once.

**Comment by jtv** [comment]
Yup, looking through the logs I see both clusters picking up the same job, so it gets processed twice.  I thought this wasn't supposed to happen when running against the same broker?

**Comment by kefin** [comment]
We have seen this multiple task execution bug with scheduled tasks -- see [#231](https://github.com/Koed00/django-q/issues/231).

**Comment by jtv** [comment]
After reading a lot of database logs, I think I found the problem!  Here's how the current code for the ORM broker grabs a task::

    if self.get_connection().filter(id=task.id, lock=task.lock).update(lock=timezone.now()):

This is quite clever.  (Which also means that it could do with some comments — cleverness leads to misunderstandings once you've forgotten the intricacies.)  It conditionally, atomically, updates Task.lock *if* nobody else has already done it.

The ORM returns the number of rows updated: 1 if the statement was able to grab the job, or 0 if it was not.  If two clusters try to do this at the same time, one will be kept waiting until the other has grabbed the job... and by that time it will simply not grab the job.

But here's what actually happens in my database (simplifying a bit for legibility)::

    UPDATE django_q_ormq
    SET lock = '2017-04-13T09:36:05.061057+00:00'::timestamptz
    WHERE id = 1

There's no check for the current Task.lock value there, and that's why two clusters can both think they grabbed the task at the same time!

And I found out why.  I'm running Django-Q 0.7.18.  Here's how its ORM broker grabs a task::

                task.lock = lock
                task.save(update_fields=['lock'])

So, the concurrency check is new, and it's not in 0.7.18 yet.  The next update should see the bug disappear!

**Comment by Eagllus** [comment]
@jtv has this problem been resolved in the newer versions?

**Comment by jtv** [comment]
Yes, thanks.  The code I showed fixed the problem — but it wasn't yet in the version I deployed originally.

The code is subtle, so it could probably do with some documentation.  Otherwise somebody might be tempted to change it later and cause a regression.
